### PR TITLE
Bring back `types.functionTo`

### PR DESCRIPTION
**Pull request by infinisil**
### Motivation for this change
We have a bunch of NixOS options that can be assigned a function, like
```nix
{
  services.hoogle.packages = p: [ p.aeson ];
}
```

These options currently all have no type specified, because there is no fitting type for that. Despite that, such options still work just fine, due to the default merging strategy being used in such a case: https://github.com/NixOS/nixpkgs/blob/f6dcf0919004123cfd9d99fc73027fa435d73ef9/lib/options.nix#L119-L123

This PR (re)introduces `types.functionTo`, which specifies that merging behavior explicitly, with more control over it. This is a cherry-pick with some fixes from https://github.com/NixOS/nixpkgs/pull/44626.

### History
- Dec 2012: @shlevy [introduces](ce6d4070a616554c0ca6948338b47cb192f7306a) `functionTo`
- Sep 2013: @oxij makes [first use](b3f4040512b360397bb8989a85776335ff3c2847) of it
- Aug 2014: @nbp [Removes the need](https://github.com/NixOS/nixpkgs/pull/3866) for @oxij's use of `functionTo` and `functionTo` as well
- Aug 2018: @basvandijk [points out](https://github.com/NixOS/nixpkgs/pull/38698#issuecomment-410894310) that we need a `types.selectorFunction` since there's a number of options needing such a type, so they open https://github.com/NixOS/nixpkgs/pull/44601 to add it
  - In the PR, @oxij [points out](https://github.com/NixOS/nixpkgs/pull/44601#issuecomment-411092134) that `selectorFunction` is just a less generic version of the previous `functionTo` and suggests readding it instead, so https://github.com/NixOS/nixpkgs/pull/44626 is created
  - @nbp Strongly opposes both PR's and [closes](https://github.com/NixOS/nixpkgs/pull/44601#issuecomment-411226577) them
- Jan 2021: @turion opens https://github.com/NixOS/nixpkgs/pull/109035, which implements such a `selectorFunction` type locally to the module, which motivates this PR

In the discussions from 2018, a number of alternatives have been brought up, but the simplicity and safety of `p: [ p.aeson ]` is just unbeatable. This has become a pattern in nixpkgs which we can't avoid that easily anymore, even if an alternative would be more suitable.

### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by roberth** [comment]
Also `allowUnfreePredicate` in `pkgs/top-level/config.nix`.

File: lib/types.nix [hunk 1]
```diff
@@ -453,6 +453,16 @@ rec {
       functor = (defaultFunctor name) // { wrapped = elemType; };
     };
 
+    functionTo = elemType: mkOptionType {
+      name = "function that evaluates to a(n) ${elemType.name}";
+      check = isFunction;
+      merge = loc: defs:
+        fnArgs: elemType.merge loc (map (fn: { inherit (fn) file; value = fn.value fnArgs; }) defs);
```
Comment (roberth) [inline lib/types.nix:460]: Do we have test cases for preserving definition locations in error messages?

**Comment by roberth** [comment]
> Now type checks the resulting function values and allows mkMerge and co.

This should be tested.

**Comment by Infinisil** [comment]
Added a bunch more tests now

File: lib/types.nix [hunk 2]
```diff
@@ -453,6 +453,16 @@ rec {
       functor = (defaultFunctor name) // { wrapped = elemType; };
     };
 
+    functionTo = elemType: mkOptionType {
+      name = "function that evaluates to a(n) ${elemType.name}";
+      check = isFunction;
+      merge = loc: defs:
+        fnArgs: (mergeDefinitions loc elemType (map (fn: { inherit (fn) file; value = fn.value fnArgs; }) defs)).mergedValue;
```
Comment (roberth) [inline lib/types.nix:460]:
```suggestion
        fnArgs: (mergeDefinitions (loc ++ ["[function body]"]) elemType (map (fn: { inherit (fn) file; value = fn.value fnArgs; }) defs)).mergedValue;
```

File: lib/tests/modules.sh [hunk 3]
```diff
@@ -262,6 +262,13 @@ checkConfigOutput true config.value.mkbefore ./types-anything/mk-mods.nix
 checkConfigOutput 1 config.value.nested.foo ./types-anything/mk-mods.nix
 checkConfigOutput baz config.value.nested.bar.baz ./types-anything/mk-mods.nix
 
+## types.functionTo
+checkConfigOutput "input is input" config.result ./functionTo/trivial.nix
+checkConfigOutput "a b" config.result ./functionTo/merging-list.nix
+checkConfigError 'A definition for option .fun. is not of type .string.. Definition values:\n- In .*wrong-type.nix' config.result ./functionTo/wrong-type.nix
```
Comment (roberth) [inline lib/tests/modules.sh:268]: Not sure if this pattern is valid. Details below.

```suggestion
checkConfigError 'A definition for option .fun.[function body]. is not of type .string.. Definition values:\n- In .*wrong-type.nix' config.result ./functionTo/wrong-type.nix
```

**Review by roberth — COMMENTED** [review]
One final touch to make the error message consistent with `listOf`. Without it, the error message is inaccurate.

**Comment by Infinisil** [comment]
Good point, now done :)

File: lib/types.nix [hunk 4]
```diff
@@ -453,6 +453,16 @@ rec {
       functor = (defaultFunctor name) // { wrapped = elemType; };
     };
 
+    functionTo = elemType: mkOptionType {
```
Comment (roberth) [inline lib/types.nix:456]: :+1: good name, leaves room for `function` with a typed input, in case that becomes relevant someday.

**Review by roberth — APPROVED** [review]
lgtm

**Comment by roberth** [comment]
The editor config check failure is for an unrelated file. Merging.

**Pull request merged**